### ISSUE TITLE
Assert being called when state machine streamhdr_handler in state NN_STREAMHDR_STATE_IDLE

### ISSUE THREAD

**Issue by bill1600** [nanomsg/nanomsg#992]
An assert is being called when the state machine streamhdr_handler 
is in the state NN_STREAMHDR_STATE_IDLE.

Here is a trace:

 2  libnanomsg.so.5.0.0!nn_err_abort [err.c : 56 + 0x5]
 3  libnanomsg.so.5.0.0!nn_streamhdr_handler [streamhdr.c : 168 + 0x42]
 4  libnanomsg.so.5.0.0!nn_fsm_feed [fsm.c : 83 + 0x6]
 5  libnanomsg.so.5.0.0!nn_fsm_event_process [fsm.c : 77 + 0xa]
 6  libnanomsg.so.5.0.0!nn_ctx_leave [ctx.c : 63 + 0x9]
 7  libnanomsg.so.5.0.0!nn_worker_routine [worker_posix.inc : 249 + 0xe]
 8  libnanomsg.so.5.0.0!nn_thread_main_routine [thread_posix.inc : 35 + 0x5]

streamhdr_handler gets in this state when you start or shutdown.

When STREAMHDR_HANDLER IS in the NN_STREAMHDR_STATE_IDLE state, 
the only valid event SRC in FSM_ACTION.

nn_leave_ctx dequeus events and runs them by calling nn_fsm_process_event.

But nn_fsm_process_event does not set event SRC to FSM_ACTION,
That happens when ou call nn_fsm_action or nn_fsm_start, 
not when you call nn_fsm_process_event.  
 
In other words, it is not valid to call nn_leave_ctx  after shutdown.  

Has anybody else seen this problem ?

**Comment by gdamore** [comment]
There are some edge cases where the code for this is not robust against certain external conditions.  Unfortunately, shutting down the library does actually trigger some of these cases.

I'd have to go through and do a more thorough analysis here, but you're deep into the state machines that are responsible for a lot of the fragility in legacy libnanomsg. 

I've transitioned almost all new development to NNG these days, which has a very very different internal architecture.

That said, if you are for some reason stuck with the legacy version of libnanomsg and need help, let me know.  We do offer paid support for this product, and I'd be happy to deep dive into it for a paying customer.

It would be helpful to see the actual value of "src" and "type" being passed in.

Alternatively, if you've done enough of the research that you understand the problem well enough to propose a fix, I'd be happy to review and (if appropriate) integrate the fix.

Failing that, I support the legacy version as I find time to do so; generally libnanomsg at this point gets reduced attention from me these days in favor of NNG.

**Comment by bill1600** [comment]
Our plan was to transition to NNG at some point.
I could keep delving into the issue on the legacy version, but maybe the smarter way of spending my time is to try out NNG immediately in compatibility mode.

**Comment by gdamore** [comment]
That would be my recommendation.
